README.md
# PowerShell module for Cloud Naming Standard

## Introduction

The CloudNaming PowerShell module can be used to programmatically generate cloud resource names for your organization.

It can be used as a standalone module, or as part of a Infrastructure as Code (IaC) pipeline.

It is highly customizable, and supports any cloud providers of your choice. The documentation and instruction can be found at the [project's wiki page](https://github.com/tyconsulting/CloudNaming-Module/wiki).

## Pipeline Examples

The [pipeline-examples](pipeline-examples) folder contains 3 sample Azure DevOps YAML pipelines to demonstrate how to use the CloudNaming module in a CI/CD pipeline:

* **[azure-pipelines-demo-1.yaml](./pipeline-examples/pipelines/azure-pipelines-demo-1.yaml)**: Beginner's level pipeline
  * Install module from a private Azure Artifacts feed
  * Does not use custom configuration file
  * Generate resource names one at a time

* **[azure-pipelines-demo-2.yaml](./pipeline-examples/pipelines/azure-pipelines-demo-2.yaml)**: Advanced level pipeline
  * Install module from PowerShell Gallery
  * Uses custom [configuration file](./pipeline-examples/config/customCloudNaming.json) located in the repository
  * Generate resource names one at a time
  * Generate multiple resource names at once

* **[azure-pipelines-install-azure-artifacts.yaml](./pipeline-examples/pipelines/azure-pipelines-install-azure-artifacts.yaml)**: Example on how to publish a customized version of the CloudNaming module to an Azure Artifacts organization-level feed.

## Additional Resources:

* Tao Yang's blog articles `Programmatically Generate Cloud Resource Names`
  * [Part 1](https://blog.tyang.org/2022/09/10/programmatically-generate-cloud-resource-names-part-1/)
  * [Part 2](https://blog.tyang.org/2022/09/10/programmatically-generate-cloud-resource-names-part-2/)

## Status

[![Status](https://github.com/tyconsulting/CloudNaming-Module/actions/workflows/main.yml/badge.svg)](https://github.com/tyconsulting/CloudNaming-Module/actions/workflows/main.yml)

CHANGELOG.md

# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased] - yyyy-mm-dd


### Added

### Changed

### Fixed

## [1.0.1] - 2022-09-14


### Added

* Links to the blog articles in the `README.md` file.

### Changed

### Fixed

* Corrected the name pattern for Azure Container Registry in the default configuration file. Original name contained disallowed character `-`. ACR names can only contain alphanumic characters.
## [1.0.0] - 2022-09-04


### Added


### Changed

* Bump the version number to 1.0.0 for initial release.

### Fixed

## [0.5.0] - 2022-09-04


### Added

* Added `newCloudNamingConfigFile` function to create new custom configuration files.

### Changed


### Fixed


## [0.4.0] - 2022-09-04


### Added



### Changed


### Fixed

* Fix parameter validation for location

## [0.4.0] - 2022-09-04


### Added



### Changed

* Changed appIdentifier to optional parameter

### Fixed


## [0.3.0] - 2022-09-03


### Added



### Changed

* Updated the name patterns for subnets in the configuration file
### Fixed


* Fixed **GetCloudNamingSupportedTypes** bug where the cloud parameter is not used

## [0.2.1] - 2022-08-29


### Added

Initial release.

### Changed


### Fixed


pipeline-examples/scripts/install-ps-module.ps1
<#
=========================================================================
AUTHOR: Tao Yang
DATE: 01/09/2022
NAME: install-ps-module.ps1
VERSION: 1.0.0
COMMENT: Install PS module from Azure Artifact feed
=========================================================================
#>
[CmdletBinding()]
param (
  [parameter(Mandatory = $false)]
  [string]$repoName = 'PSModules',

  [parameter( Mandatory = $true)]
  [string]$sourceLocation,

  [parameter( Mandatory = $true)]
  [string]$accessToken,

  [parameter( Mandatory = $true)]
  [string[]]$modules
)

#register PS repo
Write-output "Create credential for PS Repository '$repoName'."
$pw = New-Object SecureString
foreach ($char in $accessToken.ToCharArray()) {
  $pw.AppendChar($char)
}

$cred = New-Object System.Management.Automation.PSCredential 'abc', $pw

Write-Output "Register PS Repository '$repoName' with source location '$sourceLocation'."
Register-PackageSource -ProviderName 'PowerShellGet' -Name $repoName -Location $sourceLocation -Credential $cred

#install module
Foreach ($module in $modules) {
  Write-Output "Find module '$module' from '$repoName'"
  $moduleFromRepo = Find-Module -Repository $repoName -Name $module -Credential $cred
  If (!$moduleFromRepo) {
    Write-Error "Unable to find module '$module' from repo '$repoName'. Unable to continue."
    Exit -1
  } else {
    Write-output "Module '$module' found from repo '$repoName'. Latest version: '$($moduleFromrepo.version)'."
    Write-output "Installing module '$module' from repo '$repoName'..."
    Install-module -Name $module -Repository $repoName -Scope CurrentUser -force  -Credential $cred
  }

}

Write-Output "Done"

pipeline-examples/scripts/update-demo-1-parameters.ps1
#Requires -modules CloudNaming
<#
=======================================================
AUTHOR: Tao Yang
DATE: 01/09/2022
NAME: update-demo-1-parameters.ps1
VERSION: 1.0.0
COMMENT: Update the parameters of the demo-1 template
=======================================================
#>

[CmdletBinding()]
param (
  [parameter(Mandatory = $true)]
  [string]$parameterFilePath,

  [parameter(Mandatory = $false, HelpMessage = "The directory for updated parameter file")]
  [string]$updatedParameterFileArtifactName = 'UpdatedParameterFile'
)

# Import the modules
import-module CloudNaming

#variables
$appIdentifier = $env:Naming_appIdentifier
$company = $env:Naming_company
$environment = $env:Naming_environment
$location = $env:Naming_location
$vnetType = $env:Naming_vnetType
$subnetType = $env:Naming_subnetType
$nsgType = $env:Naming_nsgType
$kvType = $env:Naming_kvType
$storageAccountType = $env:Naming_storageAccountType
$resourceGroupType = $env:Naming_resourceGroupType
$privateEndpointType = $env:Naming_privateEndpointType
#Generate Names individually

#resource Group
$rgParams = @{
  cloud         = 'Azure'
  type          = $resourceGroupType
  company       = $company
  appIdentifier = $appIdentifier
  environment   = $environment
  location      = $location
  instanceCount = 1
}
Write-Verbose "Generating resource group name"
$rgName = (GetCloudResourceName @rgParams | convertfrom-Json).names[0]

#VNet
$vNetParams = @{
  cloud         = 'Azure'
  type          = $vnetType
  company       = $company
  appIdentifier = $appIdentifier
  environment   = $environment
  location      = $location
  instanceCount = 1
}
Write-Verbose "Generating virtual network name"
$vNetName = (GetCloudResourceName @vNetParams | convertfrom-Json).names[0]

#Private Endpoint subnet
$peSubnetParams = @{
  cloud         = 'Azure'
  type          = $subnetType
  workloadType  = 'pe'
  instanceCount = 1
}
Write-Verbose "Generating private endpoint subnet name"
$peSubnetName = (GetCloudResourceName @peSubnetParams | convertfrom-Json).names[0]

#Private Endpoint subnet NSG
$peSubnetNsgParams = @{
  cloud         = 'Azure'
  type          = $nsgType
  company       = $company
  appIdentifier = $appIdentifier
  environment   = $environment
  location      = $location
  instanceCount = 1
}
$peSubnetNsgName = (GetCloudResourceName @peSubnetNsgParams | convertfrom-Json).names[0]

#VM subnet
$vmSubnetParams = @{
  cloud         = 'Azure'
  type          = $subnetType
  workloadType  = 'vm'
  instanceCount = 1
}
Write-Verbose "Generating vm subnet name"
$vmSubnetName = (GetCloudResourceName @vmSubnetParams | convertfrom-Json).names[0]

#VM subnet NSG
$vmSubnetNsgParams = @{
  cloud               = 'Azure'
  type                = $nsgType
  company             = $company
  appIdentifier       = $appIdentifier
  environment         = $environment
  location            = $location
  instanceCount       = 1
  startInstanceNumber = 2
}
$vmSubnetNsgName = (GetCloudResourceName @vmSubnetNsgParams | convertfrom-Json).names[0]

#Key Vault
$kvParams = @{
  cloud         = 'Azure'
  type          = $kvType
  appIdentifier = $appIdentifier
  environment   = $environment
  location      = $location
  instanceCount = 1
}
Write-Verbose "Generatingkey vault name"
$kvName = (GetCloudResourceName @kvParams | convertfrom-Json).names[0]

#Storage Account
$storageParams = @{
  cloud         = 'Azure'
  type          = $storageAccountType
  appIdentifier = $appIdentifier
  environment   = $environment
  location      = $location
  instanceCount = 1
}
Write-Verbose "Generating storage account name"
$storageName = (GetCloudResourceName @storageParams | convertfrom-Json).names[0]

#Storage Account Private Endpoint
$storagePeParams = @{
  cloud                  = 'Azure'
  type                   = $privateEndpointType
  company                = $company
  associatedResourceType = $storageAccountType
  appIdentifier          = $appIdentifier
  instanceCount          = 1
}
Write-Verbose "Generating storage account private endpoint name"
$storagePeName = (GetCloudResourceName @storagePeParams | convertfrom-Json).names[0]

#Key Vault Private Endpoint
$kvPeParams = @{
  cloud                  = 'Azure'
  type                   = $privateEndpointType
  company                = $company
  associatedResourceType = $kvType
  appIdentifier          = $appIdentifier
  instanceCount          = 1
}
Write-Verbose "Generating key vault private endpoint name"
$kvPeName = (GetCloudResourceName @kvPeParams | convertfrom-Json).names[0]

#parse Json file
Write-Output "Parameter File Path: '$ParameterFilePath'"
$content = Get-Content -Path $ParameterFilePath -Raw
$json = ConvertFrom-Json -InputObject $content -ErrorVariable parseError
if ($parseError) {
  Throw $parseError
  exit -1
}

Write-Output "The Resource Group name is: '$rgName'"
Write-Output "The Virtual Network name is: '$vNetName'"
write-Output "The Private Endpoint subnet name is: '$peSubnetName'"
write-Output "The Private Endpoint subnet NSG name is: '$peSubnetNsgName'"
write-Output "The VM subnet name is: '$vmSubnetName'"
write-Output "The VM subnet NSG name is: '$vmSubnetNsgName'"
Write-Output "The Storage Account name is: '$storageName'"
Write-Output "The Key Vault name is: '$kvName'"
Write-Output "The Storage Account Private Endpoint name is: '$storagePeName'"
Write-Output "The Key Vault Private Endpoint name is: '$kvPeName'"


#Add names to the parameter file
Write-Output "Update the parameter file"
$json.parameters.resourceGroupName.value = $rgName
$json.parameters.vNetName.value = $vNetName
$json.parameters.storageAccountName.value = $storageName
$json.parameters.keyVaultName.value = $kvName
$json.parameters.keyVaultPrivateEndpointName.value = $kvPeName
$json.parameters.storageAccountPrivateEndpointName.value = $storagePeName
$json.parameters.peSubnetName.value = $peSubnetName
#Identify subnets
foreach ($subnet in $json.parameters.subnetConfiguration.value) {
  if ($subnet.privateEndpointNetworkPolicies -eq 'Enabled') {
    $subnet.subnetName = $peSubnetName
    $subnet.networkSecurityGroupName = $peSubnetNsgName
  } else {
    $subnet.subnetName = $vmSubnetName
    $subnet.networkSecurityGroupName = $vmSubnetNsgName
  }
}

# Save Updated Parameter File
$parameterFile = get-item $parameterFilePath
$parameterFileDir = $parameterFile.Directory
$parameterFileName = $parameterFile.Name
$UpdatedParameterFileDirPath = Join-Path $parameterFileDir $updatedParameterFileArtifactName
Write-Output "Save the updated parameter file to '$UpdatedParameterFileDirPath'"
if (!(Test-Path $UpdatedParameterFileDirPath)) {
  New-Item -ItemType Directory -Path $UpdatedParameterFileDirPath
}

$json | ConvertTo-Json -depth 100 | Out-File -FilePath $(join-path $UpdatedParameterFileDirPath $parameterFileName) -Force

pipeline-examples/publish-to-azure-artifacts/psd1-to-nuspec.ps1
#Requires -Modules PowerShellGet
#Requires -Version 5.0
<#
    =======================================================================================================
    AUTHOR:  Tao Yang
    DATE:    29/08/2022
    Version: 1.1
    Comment: generate nuget specification file (.nuspec) based on PowerShell module manifest (.psd1) file
    =======================================================================================================
#>
[CmdletBinding(PositionalBinding = $false)]
Param
(
  [Parameter(Mandatory = $true)]
  [ValidateScript({
      Test-Path $_ -PathType leaf -Include '*.psd1'
    })]
  [string]
  $ManifestPath,

  [Parameter(Mandatory = $false)]
  [ValidateScript({
      Test-Path $_ -PathType 'Container'
    })]
  [string]
  $DestinationFolder
)
#region functions
function GetEscapedString {
  [CmdletBinding()]
  [OutputType([String])]
  Param
  (
    [Parameter()]
    [string]
    $ElementValue
  )

  return [System.Security.SecurityElement]::Escape($ElementValue)
}
function GetExportedDscResources {
  [CmdletBinding(PositionalBinding = $false)]
  Param
  (
    [Parameter(Mandatory = $true)]
    [ValidateNotNullOrEmpty()]
    [PSModuleInfo]
    $PSModuleInfo
  )

  $dscResources = @()

  if (Get-Command -Name Get-DscResource -Module PSDesiredStateConfiguration -ErrorAction Ignore) {
    $OldPSModulePath = $env:PSModulePath

    try {
      $env:PSModulePath = Join-Path -Path $PSHOME -ChildPath 'Modules'
      $env:PSModulePath = "$env:PSModulePath;$(Split-Path -Path $PSModuleInfo.ModuleBase -Parent)"

      $dscResources = PSDesiredStateConfiguration\Get-DscResource -ErrorAction SilentlyContinue -WarningAction SilentlyContinue |
      Microsoft.PowerShell.Core\ForEach-Object {
        if ($_.Module -and ($_.Module.Name -eq $PSModuleInfo.Name)) {
          $_.Name
        }
      }
    } finally {
      $env:PSModulePath = $OldPSModulePath
    }
  } else {
    $dscResourcesDir = JoinPathUtility -Path $PSModuleInfo.ModuleBase -ChildPath 'DscResources' -PathType Directory
    if (Microsoft.PowerShell.Management\Test-Path $dscResourcesDir) {
      $dscResources = Microsoft.PowerShell.Management\Get-ChildItem -Path $dscResourcesDir -Directory -Name
    }
  }

  return $dscResources
}
function GetAvailableRoleCapabilityName {
  [CmdletBinding(PositionalBinding = $false)]
  Param
  (
    [Parameter(Mandatory = $true)]
    [ValidateNotNullOrEmpty()]
    [PSModuleInfo]
    $PSModuleInfo
  )

  $RoleCapabilityNames = @()

  $RoleCapabilitiesDir = JoinPathUtility -Path $PSModuleInfo.ModuleBase -ChildPath 'RoleCapabilities' -PathType Directory
  if (Microsoft.PowerShell.Management\Test-Path -Path $RoleCapabilitiesDir -PathType Container) {
    $RoleCapabilityNames = Microsoft.PowerShell.Management\Get-ChildItem -Path $RoleCapabilitiesDir `
      -Name -Filter *.psrc |
    ForEach-Object -Process {
      [System.IO.Path]::GetFileNameWithoutExtension($_)
    }
  }

  return $RoleCapabilityNames
}
function JoinPathUtility {
  <#
      .DESCRIPTION
      Utility to get the case-sensitive path, if exists.
      Otherwise, returns the output of Join-Path cmdlet.
      This is required for getting the case-sensitive paths on non-Windows platforms.
  #>
  param
  (
    [Parameter(Mandatory = $true)]
    [string]
    $Path,

    [Parameter(Mandatory = $false)]
    [string]
    $ChildPath,

    [Parameter(Mandatory = $false)]
    [string]
    [ValidateSet('File', 'Directory', 'Any')]
    $PathType = 'Any'
  )

  $JoinedPath = Microsoft.PowerShell.Management\Join-Path -Path $Path -ChildPath $ChildPath
  if (Microsoft.PowerShell.Management\Test-Path -Path $Path -PathType Container) {
    $GetChildItem_params = @{
      Path          = $Path
      ErrorAction   = 'SilentlyContinue'
      WarningAction = 'SilentlyContinue'
    }
    if ($PathType -eq 'File') {
      $GetChildItem_params['File'] = $true
    } elseif ($PathType -eq 'Directory') {
      $GetChildItem_params['Directory'] = $true
    }

    $FoundPath = Microsoft.PowerShell.Management\Get-ChildItem @GetChildItem_params |
    Where-Object -FilterScript {
      $_.Name -eq $ChildPath
    } |
    ForEach-Object -Process {
      $_.FullName
    } |
    Select-Object -First 1 -ErrorAction SilentlyContinue

    if ($FoundPath) {
      $JoinedPath = $FoundPath
    }
  }

  return $JoinedPath
}
function GetManifestHashTable {
  param
  (
    [Parameter(Mandatory = $true)]
    [ValidateNotNullOrEmpty()]
    [string]
    $Path
  )

  $Lines = $null

  try {
    $Lines = Get-Content -Path $Path -Force
  } catch {
    Throw "Unable to parse manifest file '$path'"
    Exit -1
  }

  if (-not $Lines) {
    return
  }

  $scriptBlock = [ScriptBlock]::Create( $Lines -join "`n" )

  $allowedVariables = [System.Collections.Generic.List[String]] @('PSEdition', 'PSScriptRoot')
  $allowedCommands = [System.Collections.Generic.List[String]] @()
  $allowEnvironmentVariables = $false

  try {
    $scriptBlock.CheckRestrictedLanguage($allowedCommands, $allowedVariables, $allowEnvironmentVariables)
  } catch {
    return
  }

  return $scriptBlock.InvokeReturnAsIs()
}

function NewNuSpecFile {
  [CmdletBinding(PositionalBinding = $false)]
  Param
  (
    [Parameter(Mandatory = $true, ParameterSetName = 'PublishModule')]
    [ValidateNotNullOrEmpty()]
    [string]
    $ManifestPath,

    [Parameter(Mandatory = $false)]
    [string]
    $DestinationFolder
  )

  #variables
  $Name = $null
  $Description = $null
  $Version = ''
  $Author = $null
  $CompanyName = $null
  $Copyright = $null
  $requireLicenseAcceptance = 'false'

  $PSModuleInfo = Test-ModuleManifest -Path $ManifestPath
  If (-not $PSModuleInfo) {
    Throw "Failed to retrieve module information from manifest '$ManifestPath'"
    exit -1
  }

  $Name = $PSModuleInfo.Name
  $Description = $PSModuleInfo.Description
  $Version = $PSModuleInfo.Version
  $Author = $PSModuleInfo.Author
  $CompanyName = $PSModuleInfo.CompanyName
  $Copyright = $PSModuleInfo.Copyright

  If (-not $PSBoundParameters.ContainsKey('DestinationFolder')) {
    $DestinationFolder = $PSModuleInfo.ModuleBase
  }
  $NuspecPath = Microsoft.PowerShell.Management\Join-Path -Path $DestinationFolder -ChildPath "$($PSModuleInfo.Name).nuspec"

  if ($PSModuleInfo.PrivateData -and
    ($PSModuleInfo.PrivateData.GetType().ToString() -eq 'System.Collections.Hashtable') -and
    $PSModuleInfo.PrivateData['PSData'] -and
    ($PSModuleInfo.PrivateData['PSData'].GetType().ToString() -eq 'System.Collections.Hashtable')
  ) {
    if ($PSModuleInfo.PrivateData.PSData['Tags']) {
      $Tags = $PSModuleInfo.PrivateData.PSData.Tags
    }

    if ($PSModuleInfo.PrivateData.PSData['ReleaseNotes']) {
      $ReleaseNotes = $PSModuleInfo.PrivateData.PSData.ReleaseNotes
    }

    if ($PSModuleInfo.PrivateData.PSData['LicenseUri']) {
      $LicenseUri = $PSModuleInfo.PrivateData.PSData.LicenseUri
    }

    if ($PSModuleInfo.PrivateData.PSData['IconUri']) {
      $IconUri = $PSModuleInfo.PrivateData.PSData.IconUri
    }

    if ($PSModuleInfo.PrivateData.PSData['ProjectUri']) {
      $ProjectUri = $PSModuleInfo.PrivateData.PSData.ProjectUri
    }

    if ($PSModuleInfo.PrivateData.PSData['Prerelease']) {
      $psmoduleInfoPrereleaseString = $PSModuleInfo.PrivateData.PSData.Prerelease
      if ($psmoduleInfoPrereleaseString -and $psmoduleInfoPrereleaseString.StartsWith('-')) {
        $Version = [string]$Version + $psmoduleInfoPrereleaseString
      } else {
        $Version = [string]$Version + '-' + $psmoduleInfoPrereleaseString
      }
    }

    if ($PSModuleInfo.PrivateData.PSData['RequireLicenseAcceptance']) {
      $requireLicenseAcceptance = $PSModuleInfo.PrivateData.PSData.requireLicenseAcceptance.ToString().ToLower()
      if ($requireLicenseAcceptance -eq 'true') {
        if (-not $LicenseUri) {
          $message = "'LicenseUri' is not specified. 'LicenseUri' must be provided when user license acceptance is required."
          Throw $message
        }

        $LicenseFilePath = JoinPathUtility -Path $PSModuleInfo.ModuleBase -ChildPath 'License.txt' -PathType File
        if (-not $LicenseFilePath -or -not (Test-Path -Path $LicenseFilePath -PathType Leaf)) {
          $message = 'License.txt not Found. License.txt must be provided when user license acceptance is required.'
          Throw $message
        }

        if ($null -eq (Get-Content -LiteralPath $LicenseFilePath)) {
          $message = 'License.txt is empty.'
          Throw $message
        }
      } elseif ($requireLicenseAcceptance -ne 'false') {
        $InvalidValueForRequireLicenseAcceptance = "The specified value '{0}' for the parameter '{1}' is invalid. It should be $true or $false." -f ($requireLicenseAcceptance, 'requireLicenseAcceptance')
        Write-Warning -Message $InvalidValueForRequireLicenseAcceptance
      }
    }
  }

  # Add PSModule and PSGet format version tags
  if (-not $Tags) {
    $Tags = @()
  }


  $DependentModuleDetails = @()
  $Tags += 'PSModule'
  $ModuleManifestHashTable = GetManifestHashTable -Path $ManifestPath
  if ($PSModuleInfo.ExportedCommands.Count) {
    if ($PSModuleInfo.ExportedCmdlets.Count) {
      $Tags += 'PSIncludes_Cmdlet'
      $Tags += $PSModuleInfo.ExportedCmdlets.Keys | Microsoft.PowerShell.Core\ForEach-Object {
        "PSCmdlet_$_"
      }
    }

    if ($PSModuleInfo.ExportedFunctions.Count) {
      $Tags += 'PSIncludes_Function'
      $Tags += $PSModuleInfo.ExportedFunctions.Keys | Microsoft.PowerShell.Core\ForEach-Object {
        "PSFunction_$_"
      }
    }

    $Tags += $PSModuleInfo.ExportedCommands.Keys | Microsoft.PowerShell.Core\ForEach-Object {
      "PSCommand_$_"
    }
  }

  $dscResourceNames = GetExportedDscResources -PSModuleInfo $PSModuleInfo
  if ($dscResourceNames) {
    $Tags += 'PSIncludes_DscResource'

    $Tags += $dscResourceNames | Microsoft.PowerShell.Core\ForEach-Object {
      "PSDscResource_$_"
    }
  }

  $RoleCapabilityNames = GetAvailableRoleCapabilityName -PSModuleInfo $PSModuleInfo
  if ($RoleCapabilityNames) {
    $Tags += 'PSIncludes_RoleCapability'

    $Tags += $RoleCapabilityNames | Microsoft.PowerShell.Core\ForEach-Object {
      "PSRoleCapability_$_"
    }
  }

  # Populate the module dependencies elements from RequiredModules and
  # NestedModules properties of the current PSModuleInfo
  $DependentModuleDetails = @()
  $requiredModules = @()
  #$requiredModules += $PSModuleInfo.RequiredModules
  #$requiredModules += $PSModuleInfo.NestedModules
  if ($ModuleManifestHashTable.RequiredModules) {
    $requiredModules += $ModuleManifestHashTable.RequiredModules
  }
  if ($ModuleManifestHashTable.NestedModules) {
    $requiredModules += $ModuleManifestHashTable.NestedModules
  }

  Write-Verbose "Total dependent modules: $($requiredModules.count)"
  Foreach ($requiredModule in $requiredModules) {
    $DependentModuleDetail = @{}
    if ($requiredModule.GetType().ToString() -eq 'System.Collections.Hashtable') {
      $ModuleName = $requiredModule.ModuleName
      Write-Verbose "Processing dependency module '$ModuleName'"
      if ($requiredModule.Keys -Contains 'RequiredVersion') {
        Write-Verbose "'$ModuleName': Required version: $($requiredModule.RequiredVersion)"
        $DependentModuleDetail.add('RequiredVersion', $requiredModule.RequiredVersion)
      } elseif ($requiredModule.Keys -Contains 'ModuleVersion') {
        Write-Verbose "$ModuleName': Module version: $($requiredModule.ModuleVersion)"
        $DependentModuleDetail.add('ModuleVersion', $requiredModule.ModuleVersion)
      }
    } else {
      # Just module name was specified
      Write-Verbose "$ModuleName': Module version not specified."
      $ModuleName = $requiredModule.ToString()
    }
    $DependentModuleDetail.add('Name', $ModuleName)
    $DependentModuleDetails += $DependentModuleDetail
  }

  $dependencies = @()
  ForEach ($Dependency in $DependentModuleDetails) {
    $ModuleName = $Dependency.Name
    $VersionString = $null

    # Version format in NuSpec:
    # "[2.0]" --> (== 2.0) Required Version
    # "2.0" --> (>= 2.0) Minimum Version
    #
    # When only MaximumVersion is specified in the ModuleSpecification
    # (,1.0]  = x <= 1.0
    #
    # When both Minimum and Maximum versions are specified in the ModuleSpecification
    # [1.0,2.0] = 1.0 <= x <= 2.0

    if ($Dependency.Keys -Contains 'RequiredVersion') {
      $VersionString = "[$($Dependency.RequiredVersion)]"
    } elseif ($Dependency.Keys -Contains 'ModuleVersion') {
      $VersionString = "$($Dependency.ModuleVersion)"
    }

    if ([System.string]::IsNullOrWhiteSpace($VersionString)) {
      $dependencies += "<dependency id='$($ModuleName)'/>"
    } else {
      $dependencies += "<dependency id='$($ModuleName)' version='$($VersionString)' />"
    }
  }

  # Populate the nuspec elements
  $nuspec = @"
<?xml version="1.0"?>
<package >
    <metadata>
        <id>$(GetEscapedString -ElementValue "$Name")</id>
        <version>$($Version)</version>
        <authors>$(GetEscapedString -ElementValue "$Author")</authors>
        <owners>$(GetEscapedString -ElementValue "$CompanyName")</owners>
        <description>$(GetEscapedString -ElementValue "$Description")</description>
        <releaseNotes>$(GetEscapedString -ElementValue "$ReleaseNotes")</releaseNotes>
        <requireLicenseAcceptance>$($requireLicenseAcceptance.ToString())</requireLicenseAcceptance>
        <copyright>$(GetEscapedString -ElementValue "$Copyright")</copyright>
        <tags>$(if($Tags){ GetEscapedString -ElementValue ($Tags -join ' ')})</tags>
        $(if($LicenseUri)
{
        "<licenseUrl>$(GetEscapedString -ElementValue "$LicenseUri")</licenseUrl>"
})
        $(if($ProjectUri)
{
        "<projectUrl>$(GetEscapedString -ElementValue "$ProjectUri")</projectUrl>"
})
        $(if($IconUri)
{
        "<iconUrl>$(GetEscapedString -ElementValue "$IconUri")</iconUrl>"
})
        <dependencies>
            $dependencies
        </dependencies>
    </metadata>
</package>
"@

  try {
    # Remove existing nuspec file

    if ($NuspecPath -and (Test-Path -Path $NuspecPath -PathType Leaf)) {
      Write-Warning "Nuspec file '$NuspecPath' already exists. It will be overwritten."
      Microsoft.PowerShell.Management\Remove-Item $NuspecPath -Force -ErrorAction SilentlyContinue -WarningAction SilentlyContinue -Confirm:$false -WhatIf:$false
    }

    Microsoft.PowerShell.Management\Set-Content -Value $nuspec -Path $NuspecPath -Force -Confirm:$false -WhatIf:$false

    if ($LASTEXITCODE -or -not $NuspecPath -or -not (Test-Path -Path $NuspecPath -PathType Leaf)) {
      $message = "failed to create nuspec file '$NuspecPath'"
      $errorId = 'FailedToCreateNuspecFile'

      Write-Error -Message $message -ErrorId $errorId -Category InvalidOperation
      return
    }
  } finally {
    if ($NuspecPath -and (Test-Path -Path $NuspecPath -PathType Leaf) -and (Get-Content -LiteralPath $NuspecPath -Raw)) {
      Write-Output -InputObject $NuspecPath
    }
  }
}
#endregion

#region main
Write-Output -InputObject "Generating .nuspec file based on PowerShell Module Manifest '$ManifestPath'"
$param = @{
  'ManifestPath' = $ManifestPath
}
If ($PSBoundParameters.ContainsKey('DestinationFolder')) {
  $param.Add('DestinationFolder', $DestinationFolder)
}
$NuspecFile = NewNuSpecFile @param
If ($NuspecFile) {
  Write-Output "Nuspec file created - '$NuspecFile'."
  Write-Output "Done. "
} else {
  Write-Error "Failed to create Nuspec file."
}
#endregion